### PR TITLE
feat: Translation

### PR DESCRIPTION
**Pull request by claudiahub**
This Poc enables support on SCDF for multi-language translations, dynamically at runtime.

- Easily maintainable : JSON files for each supported language.

- There is no need to do extra builds when adding new translations

- Language dropdown switcher available on "Settings" modal

### CODE REVIEW

**Comment by edy-ap** [comment]
Excellent work!! !! I hope it will be integrated in the next version.

File: ui/src/assets/i18n/de.json [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        "startersForRabbitMQDocker": "Stream-Anwendungsstarter für RabbitMQ/Docker",
+        "startersForMaven": "Task-Anwendungsstarter für Maven",
+        "startersForDocker": "Task-Anwendungsstarter für Docker",
+        "forceImport": "Erzwingen, die Anwendungen werden importiert und installiert, auch wenn sie bereits vorhanden sind, jedoch nur, wenn sie noch nicht verwendet werden",
+        "importApplications": "Anwendung(en) importieren",
+        "importingApplications": "Anwendung(en) werden importiert..."
+      }
+    },
+    "unregister": {
+      "confirmUnregisterApplication": "Anmeldung abmelden bestätigen",
```
Comment (klopfdreh) [inline ui/src/assets/i18n/de.json:74]: Just a small hint Application(s) translates to Anwendung(en) - Anmeldung translates to Login.

**Comment by klopfdreh** [comment]
Should I close https://github.com/spring-cloud/spring-cloud-dataflow-ui/pull/1787 in favor of this PR?

**Comment by oodamien** [comment]
Hi @klopfdreh, 
We are working on complete this PR. I will postpone your work (german translation) a soon as it's complete. 
Then, if you have time to review the german, it will be awesome.

**Comment by klopfdreh** [comment]
All right! 👍 Ping me again if I should review the translation.

File: ui/src/app/tasks-jobs/schedules/destroy/destroy.component.html [hunk 2]
```diff
@@ -1,25 +1,17 @@
...
-    listed below.<br />Are you sure?
-    <br />
+    <div [innerHTML]="'schedules.destroy.description2' | translate: {count: schedules.length}"></div>
     <table class="table table-striped">
       <thead>
         <tr>
-          <th class="left">Name</th>
-          <th class="left">Task name</th>
-          <th class="left">Platform</th>
+          <th class="left">{{ 'schedules.main.name' | translate }}</th>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/schedules/destroy/destroy.component.html:12]: Texts like `Name` and `Platform` are shown on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/schedules/schedule/schedule.component.html [hunk 3]
```diff
@@ -1,34 +1,43 @@
...
+        [titleModal]="'schedules.schedule.information' | translate"
+        keyContext="schedule"
+        name="{{ schedule.name }}"
+        id="info"
+      >
         <ng-template>
           <div class="block card-block-keyvalue">
             <div class="row">
-              <div class="key">Name</div>
+              <div class="key">{{ 'schedules.main.name' | translate }}</div>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/schedules/schedule/schedule.component.html:26]: Texts like `Name`, `Platform`, `Status` and `Definition` are visibile on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/schedules/schedules.component.html [hunk 4]
```diff
@@ -63,14 +69,20 @@ <h1>Schedules</h1>
     <clr-dg-cell>{{ schedule.platform || 'N/A' }}</clr-dg-cell>
     <clr-dg-cell>{{ schedule.cronExpression || 'N/A' }}</clr-dg-cell>
     <clr-dg-action-overflow *ngIf="!grouped">
-      <button class="action-item" (click)="details(schedule)">Details</button>
-      <button class="action-item" (click)="taskDetails(schedule)">Task details</button>
-      <button class="action-item" (click)="createSchedule(schedule)">Add new schedule</button>
-      <button class="action-item" (click)="destroySchedules([schedule])">Destroy</button>
+      <button class="action-item" (click)="details(schedule)">{{ 'schedules.main.details' | translate }}</button>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/schedules/schedules.component.html:72]: Texts like `Name`, `Platform` and `Details` are visibile on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/clone/clone.component.html [hunk 5]
```diff
@@ -10,27 +10,27 @@ <h3 *ngIf="tasks?.length > 1" class="modal-title">Clone Tasks</h3>
...
-            <div class="alert-text">Duplicate task name, please check input names.</div>
+            <div class="alert-text">{{ 'tasks.clone.duplication' | translate }}</div>
           </div>
         </div>
       </div>
       <div *ngFor="let task of tasks; let i = index" class="card" style="padding: 10px 0">
         <div class="card-block">
           <div class="clr-form-control clr-row" style="margin-top: 0; height: 42px">
-            <label class="clr-control-label clr-col-12 clr-col-md-2">Definition:</label>
+            <label class="clr-control-label clr-col-12 clr-col-md-2">{{ 'tasks.main.definition' | translate }}:</label>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/clone/clone.component.html:27]: Texts like `Name` and `Definition` are visibile on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/create/create.component.html [hunk 6]
```diff
@@ -25,31 +29,33 @@ <h3 class="modal-title">Create task</h3>
...
               <clr-icon class="alert-icon" shape="exclamation-circle"></clr-icon>
             </div>
-            <div class="alert-text">Duplicate task name, please check input names.</div>
+            <div class="alert-text">{{ 'tasks.create.modal.duplicateName' | translate }}</div>
           </div>
         </div>
       </div>
       <div class="clr-form-control clr-row" style="margin-top: 0; max-height: 16px">
-        <label class="clr-control-label clr-col-12 clr-col-md-2">Definition</label>
+        <label class="clr-control-label clr-col-12 clr-col-md-2">{{ 'tasks.main.definition' | translate }}</label>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/create/create.component.html:46]: Texts like `Definition`, `Name` and `Task Name` are visible on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/destroy/destroy.component.html [hunk 7]
```diff
@@ -1,24 +1,16 @@
...
-    >
-    listed below.<br />Are you sure?
-    <br />
+    <div [innerHTML]="'tasks.destroy.content2' | translate: {count: tasks?.length}"></div>
     <table class="table table-striped">
       <thead>
         <tr>
-          <th class="left">Name</th>
-          <th class="left">Type</th>
+          <th class="left">{{ 'tasks.main.name' | translate }}</th>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/destroy/destroy.component.html:12]: Texts like `Name` and `Type` are visible on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/launch/builder/builder.component.html [hunk 8]
```diff
@@ -31,7 +33,7 @@
             <div class="line-toggle">
               <a class="toggle" (click)="state.platform = !state.platform">
                 <clr-icon shape="angle" [class.active]="state.platform"></clr-icon>
-                Platform
+                {{ 'tasks.launch.builder.platform' | translate }}
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/launch/builder/builder.component.html:36]: `Platform` text visible on other pages.
Can we move this to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/launch/builder/builder.component.html [hunk 9]
```diff
@@ -140,10 +142,10 @@
                   </clr-tooltip>
                 </div>
 
-                <div class="form-control form-control-label">Version</div>
+                <div class="form-control form-control-label">{{ 'tasks.launch.builder.version' | translate }}</div>
               </div>
               <div class="line-body">
-                <div class="form-control form-control-label">Properties</div>
+                <div class="form-control form-control-label">{{ 'tasks.launch.builder.properties' | translate }}</div>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/launch/builder/builder.component.html:148]: Texts like `Version` and `Properties` are visible on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/launch/free-text/free-text.component.html [hunk 10]
```diff
@@ -28,23 +28,23 @@
...
         </clr-tooltip-content>
       </clr-tooltip>
 
-      <button id="btn-export" type="button" class="btn btn-sm btn btn-secondary" (click)="exportProps()">Export</button>
+      <button id="btn-export" type="button" class="btn btn-sm btn btn-secondary" (click)="exportProps()">
+        {{ 'tasks.launch.main.export' | translate }}
+      </button>
       <button id="btn-copy" type="button" class="btn btn-sm btn btn-secondary" (click)="copyPropsToClipboard()">
-        Copy to Clipboard
+        {{ 'tasks.launch.main.copy' | translate }}
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/launch/free-text/free-text.component.html:42]: `Copy to Clipboard` is rendered on other sections.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/task/task.component.html [hunk 11]
```diff
@@ -16,40 +16,45 @@ <h1 class="page-title">
...
             <div class="row">
-              <div class="key">Status</div>
+              <div class="key">{{ 'tasks.main.status' | translate }}</div>
               <div class="value">
                 <span class="{{ task.labelStatusClass() }}">{{ task.status }}</span>
               </div>
             </div>
             <div class="row">
-              <div class="key">Applications</div>
+              <div class="key">{{ 'tasks.task.applications' | translate }}</div>
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/task/task.component.html:57]: Texts like `Applications`, `Description`, `Status` and `Definition` are visible on other pages.
Can we move these to `commons` translations ?

File: ui/src/app/tasks-jobs/tasks/tasks.component.html [hunk 12]
```diff
@@ -1,11 +1,15 @@
 <h1 class="page-title-actions">
-  Tasks
+  {{ 'tasks.main.title' | translate }}
   <span class="actions">
     <button type="button" class="btn btn-sm btn-primary" (click)="createTask()" [appRole]="['ROLE_CREATE']">
-      Create task
+      {{ 'tasks.main.create' | translate }}
+    </button>
+    <button type="button" class="btn btn-sm btn-secondary" grafanaDashboardTasks>
+      {{ 'tasks.main.grafana' | translate }}
```
Comment (claudiahub) [inline ui/src/app/tasks-jobs/tasks/tasks.component.html:8]: Texts like `Grafana Dashboard` and `Wavefront Dashboard` are visible on other pages.
Can we move these to `commons` translations ?

File: ui/src/assets/i18n/en.json [hunk 13]
```diff
@@ -0,0 +1,1100 @@
...
+      "argumentsDesc": "Arguments can be prefixed by <code>--</code> (optional). Example:",
+      "invalidArguments": "One or more arguments are invalid.<br />Example: <code>myarg=myvalue</code>.",
+      "properties": "Properties",
+      "propertiesDesc": "Parameters should start with an <code>app.</code> or <code>scheduler.</code>. Example:",
+      "invalidProperties": "One or more parameters are invalid.<br />Example: <code>app.myarg=myvalue</code>.",
+      "create": "Create schedule(s)",
+      "creating": "Creating schedule(s)",
+      "message": {
+        "groupSeclectionNotFoundTitle": "Group selection not found.",
+        "scheculeSuccessTitle": "Schedule creation",
```
Comment (claudiahub) [inline ui/src/assets/i18n/en.json:1050]: Is `schecule` a typo for `schedule` ?

**Review by claudiahub — COMMENTED** [review]
Excellent job @oodamien on adding all of these translations.
I've commented on some labels that can be grouped into `commons` and reused through all the UI. 
Rest of PR looks good to me.

File: ui/src/app/streams/streams/create/create.component.ts [hunk 14]
```diff
@@ -205,11 +204,17 @@ export class CreateComponent implements OnInit {
 
   createStream(): void {
     if (!this.flo.dsl || !this.flo.dsl.trim()) {
-      this.notificationService.error('Invalid stream(s)', 'Please, enter one or more valid streams.');
+      this.notificationService.error(
+        this.translate.instant('streams.create.modal.message.invalidStreamTitle'),
```
[the remark this answers is not in the document]
Comment (claudiahub) [reply ui/src/app/streams/streams/create/create.component.ts:208]: Titles are slightly different, we keep both versions.

**Comment by klopfdreh** [comment]
Hey @oodamien and @claudiahub - we are currently testing the ui with the 2.9.4 server. So far it is working like a charm and the switch between languages on the fly looks impressive. 😄

**Pull request merged**